From f3e682f1a0ac5fed7264248ad3c33a1f911c850e Mon Sep 17 00:00:00 2001
From: Muhammad Usama Anjum <usama.anjum@collabora.com>
Date: Thu, 16 Oct 2025 15:08:04 +0500
Subject: [PATCH 037/188] ACPI: button: Cancel hibernation if button is pressed
 during hibernation

acpi_pm_wakeup_event() is called from acpi_button_notify() which is
called when power button is pressed. The system is worken up from s2idle
in this case by setting hard parameter to pm_wakeup_dev_event().

Call acpi_pm_wakeup_event() if power button is pressed and hibernation
is in progress. Set the hard parameter such that pm_system_wakeup()
gets called which increments pm_abort_suspend counter. The explicit call
to acpi_pm_wakeup_event() is necessary as ACPI button device has the
wakeup source. Hence call to input_report_key() with input device
doesn't call pm_system_wakeup() as it doesn't have wakeup source
registered.

Hence hibernation would be cancelled as in hibernation path, this counter
is checked if it should be aborted.

Signed-off-by: Muhammad Usama Anjum <usama.anjum@collabora.com>
---
 drivers/acpi/button.c    | 12 ++++++++----
 drivers/acpi/device_pm.c |  6 ++++++
 include/acpi/acpi_bus.h  |  4 ++++
 3 files changed, 18 insertions(+), 4 deletions(-)

diff --git a/drivers/acpi/button.c b/drivers/acpi/button.c
index 3836ee75d..9d174cf23 100644
--- a/drivers/acpi/button.c
+++ b/drivers/acpi/button.c
@@ -21,6 +21,7 @@
 #include <linux/dmi.h>
 #include <linux/platform_device.h>
 #include <acpi/button.h>
+#include <linux/suspend.h>
 
 #define ACPI_BUTTON_CLASS		"button"
 #define ACPI_BUTTON_FILE_STATE		"state"
@@ -451,6 +452,7 @@ static void acpi_button_notify(acpi_handle handle, u32 event, void *data)
 	struct acpi_device *device = button->adev;
 	struct input_dev *input;
 	int keycode;
+	bool hard = false;
 
 	switch (event) {
 	case ACPI_BUTTON_NOTIFY_STATUS:
@@ -463,14 +465,16 @@ static void acpi_button_notify(acpi_handle handle, u32 event, void *data)
 		return;
 	}
 
-	acpi_pm_wakeup_event(button->dev);
+	input = button->input;
+	keycode = test_bit(KEY_SLEEP, input->keybit) ? KEY_SLEEP : KEY_POWER;
+
+	if (event == ACPI_BUTTON_NOTIFY_STATUS)
+		hard = pm_sleep_transition_in_progress();
+	acpi_pm_wakeup_event_hard(button->dev, hard);
 
 	if (button->suspended || event == ACPI_BUTTON_NOTIFY_WAKE)
 		return;
 
-	input = button->input;
-	keycode = test_bit(KEY_SLEEP, input->keybit) ? KEY_SLEEP : KEY_POWER;
-
 	input_report_key(input, keycode, 1);
 	input_sync(input);
 	input_report_key(input, keycode, 0);
diff --git a/drivers/acpi/device_pm.c b/drivers/acpi/device_pm.c
index aa55ecfc2..5aa10d55a 100644
--- a/drivers/acpi/device_pm.c
+++ b/drivers/acpi/device_pm.c
@@ -526,6 +526,12 @@ void acpi_pm_wakeup_event(struct device *dev)
 }
 EXPORT_SYMBOL_GPL(acpi_pm_wakeup_event);
 
+void acpi_pm_wakeup_event_hard(struct device *dev, bool hard)
+{
+	pm_wakeup_dev_event(dev, 0, acpi_s2idle_wakeup() || hard);
+}
+EXPORT_SYMBOL_GPL(acpi_pm_wakeup_event_hard);
+
 static void acpi_pm_notify_handler(acpi_handle handle, u32 val, void *not_used)
 {
 	struct acpi_device *adev;
diff --git a/include/acpi/acpi_bus.h b/include/acpi/acpi_bus.h
index 714d111d8..9e5b2e64b 100644
--- a/include/acpi/acpi_bus.h
+++ b/include/acpi/acpi_bus.h
@@ -806,6 +806,7 @@ static inline bool acpi_quirk_skip_gpio_event_handlers(void)
 
 #ifdef CONFIG_PM
 void acpi_pm_wakeup_event(struct device *dev);
+void acpi_pm_wakeup_event_hard(struct device *dev, bool hard);
 acpi_status acpi_add_pm_notifier(struct acpi_device *adev, struct device *dev,
 			void (*func)(struct acpi_device_wakeup_context *context));
 acpi_status acpi_remove_pm_notifier(struct acpi_device *adev);
@@ -816,6 +817,9 @@ int acpi_pm_set_device_wakeup(struct device *dev, bool enable);
 static inline void acpi_pm_wakeup_event(struct device *dev)
 {
 }
+static inline void acpi_pm_wakeup_event_hard(struct device *dev, bool hard)
+{
+}
 static inline acpi_status acpi_add_pm_notifier(struct acpi_device *adev,
 					       struct device *dev,
 					       void (*func)(struct acpi_device_wakeup_context *context))
-- 
2.55.0

